crates/latch_core/src/ecs/system_macros.rs

```diff
@@ -0,0 +1,181 @@
+// system_macros.rs - Ergonomic macros for ECS system definition
+//
+// These macros provide a clean, user-friendly API for defining systems
+// that iterate over entities with specific components, hiding the
+// complexity of archetype iteration and double-buffering.
+
+/// Execute a closure for each entity with the specified components.
+/// 
+/// This macro provides a clean API for iterating over entities without
+/// exposing archetype implementation details. It automatically handles:
+/// - Finding all archetypes with the required components
+/// - Extracting component slices using double-buffering
+/// - Iterating over entities in parallel (if using rayon)
+/// 
+/// The closure receives pairs of `(&current, &mut next)` references for each component:
+/// - `current`: Read from the stable state (previous tick)
+/// - `next`: Write to the new state (next tick)
+/// 
+/// # Example
+/// 
+/// ```ignore
+/// use latch_core::for_each_entity;
+/// 
+/// // Simple physics update
+/// for_each_entity!(world, [Position, Velocity], |
+///     (pos_curr, pos_next),
+///     (vel_curr, vel_next)
+/// | {
+///     pos_next.x = pos_curr.x + vel_curr.x as i32;
+///     pos_next.y = pos_curr.y + vel_curr.y as i32;
+///     *vel_next = *vel_curr;
+/// });
+/// ```
+/// 
+/// # Parallel Iteration
+/// 
+/// For parallel iteration with rayon, use the `parallel` variant:
+/// 
+/// ```ignore
+/// use rayon::prelude::*;
+/// 
+/// for_each_entity!(world, [Position, Velocity], |
+///     (pos_curr, pos_next),
+///     (vel_curr, vel_next)
+/// | {
+///     // This closure will be called for each entity in parallel
+///     pos_next.x = pos_curr.x + vel_curr.x as i32;
+///     pos_next.y = pos_curr.y + vel_curr.y as i32;
+/// }, parallel);
+/// ```
+#[macro_export]
+macro_rules! for_each_entity {
+    // Sequential iteration: for_each_entity!(world, [C1, C2, ...], |args| { body })
+    ($world:expr, [$($Component:ty),+ $(,)?], |$(($curr:ident, $next:ident)),+ $(,)?| $body:expr) => {{
```

Comment:
We shouldn't give the developers the option of serial versus parallel. Our storage system is designed to make parallel safe, so we ALWAYS do parallel under the hood and hide this implementation detail.